Add Convolutional1dComponent in nnet3 and its test code

freewym opened the pull request:
It passed the test code.

danpovey commented on src/nnet3/nnet-test-utils.cc:
```diff
@@ -804,6 +804,18 @@ static void GenerateRandomComponentConfig(std::string *component_type,
     }
+    case 17: {
+      *component_type = "Convolutional1dComponent";
+      int32 patch_stride = 10, patch_step = 1, patch_dim = 4;
```
I think it would be better to test a variety of values here- this may be masking bugs.

danpovey commented on src/nnet3/nnet-simple-component.h:
```diff
@@ -891,6 +891,131 @@ class NaturalGradientPerElementScaleComponent: public PerElementScaleComponent {
+  virtual std::string Type() const { return "Convolutional1dComponent"; }
+  virtual int32 Properties() const {
+    return kSimpleComponent|kUpdatableComponent|kBackpropNeedsInput;
+	   // |kBackpropAdds;
```
Remove this comment.

danpovey commented on src/nnet3/nnet-simple-component.h:
```diff
@@ -891,6 +891,131 @@ class NaturalGradientPerElementScaleComponent: public PerElementScaleComponent {
+  }
+
+  virtual void Propagate(const ComponentPrecomputedIndexes *indexes,
+                 const CuMatrixBase<BaseFloat> &in,
```
fix indentation

danpovey commented on src/nnet3/nnet-simple-component.cc:
```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
+    RearrangeIndexes(reversed_column_map, &rearranged_column_map);
+    for (int32 p = 0; p < rearranged_column_map.size(); p++) {
+      CuArray<int32> cu_cols(rearranged_column_map[p]);
+      in_deriv->AddCols(patches_deriv, cu_cols);
```
I think maybe this should be CopyCols.  Actually I'm very surprised that the tests passed because this component seems to add to its input and never zero it, yet the properties says that it just sets its input. TestSimpleComponentPropagateProperties() should have failed.  [see output_deriv2].

danpovey replied:
OK, I see now that AddCols is appropriate, because you're calling it multiple times.  But I don't see that you ever zero in_deriv.  Ooh.. I see now.

freewym replied:
It's at the first line of Backprop when doing resizing

danpovey commented on src/nnet3/nnet-simple-component.cc:
```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
+                                        Component *to_update_in,
+                                        CuMatrixBase<BaseFloat> *in_deriv) const {
+  if (in_deriv)
+    static_cast<CuMatrix<BaseFloat>* >(in_deriv)->Resize(out_deriv.NumRows(), InputDim());
```
Ouch!!  You should never static_cast CuMatrixBase to CuMatrix like this!!!  How could this code possibly work if it really pointed to a sub-matrix?  And it will, if it's ever used in a real neural net.    I hope you didn't do this in any previous code.
The code requires that in_deriv already have the correct size.  If you want to zero it you can call SetZero().
But actually, don't call SetZero().  The framework is designed so that if it's more natural to add to the derivative, you just set the flag kBackpropAdds, and it's then the responsibility of the calling code to zero the derivative (if needed).  That way, if the derivative needs to be added to something else, you may be able to avoid redundant copies of matrices.  So add the flag kBackpropAdds, and remove this statement.

freewym replied:
Sorry about that. I did the static_cast because  Backprop() of CNN in nnet2 calls Resize() at the beginning. I will fix it.

freewym commented:
I fixed the those bugs. Also I removed the call of Resize() at the beginnig of Convolutional1dComponent::Backprop() in nnet2 and changed a line in its nnet-component-test.cc accordingly. Please let me know if there are any problems.
.
Yiming

freewym commented on src/nnet3/nnet-test-utils.cc:
```diff
@@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
+	    patch_dim = 4 + Rand () % 5;
+
+      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
+      while ((patch_stride - patch_dim) % patch_step != 0)
```
Sorry. no while needed. will replace it with a single expression.

danpovey replied:
In nnet2, in_deriv is supposed to be resized by the Backprop routine, that
it why it is typed CuMatrix not CuMatrixBase.  In nnet2, it's different.


On Sun, Sep 20, 2015 at 5:34 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/156#discussion_r39935031>:
>
> > @@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >        os << "input-dim=" << input_dim << " output-dim=" << output_dim;
> >        break;
> >      }
> > +    case 17: {
> > +      *component_type = "Convolutional1dComponent";
> > +      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4,
> > +	    patch_dim = 4 + Rand () % 5;
> > +
> > +      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
> > +      while ((patch_stride - patch_dim) % patch_step != 0)
>
> Sorry. no while needed. will replace it with a single expression.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/156/files#r39935031>.
>

freewym replied:
Restored.

danpovey replied:
OK.  Please rebase your changes so that only the initial commit remains.
Dan


On Sun, Sep 20, 2015 at 5:49 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/156#discussion_r39935212>:
>
> > @@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >        os << "input-dim=" << input_dim << " output-dim=" << output_dim;
> >        break;
> >      }
> > +    case 17: {
> > +      *component_type = "Convolutional1dComponent";
> > +      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4,
> > +	    patch_dim = 4 + Rand () % 5;
> > +
> > +      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
> > +      while ((patch_stride - patch_dim) % patch_step != 0)
>
> Restored.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/156/files#r39935212>.
>

freewym replied:
Done. One issue: I left the previous 2 commits (by david and vijay) in the interactive editor when using rebase, so currently there are 3 commits in this pull request. But actually I didn't modify anything in their commits. I am not sure if it is fine for you to merge only mine. If not, I can create a new pull request. Sorry about it.
Yiming

danpovey replied:
You must have done something wrong-- those shouldn't appear there.  See if
you can find someone to help you fix it.  To-morrow is fine.
Meanwhile, you could also try to add a test example in nnet-test-utils.cc
that tests the convolutional component.
Dan


On Sun, Sep 20, 2015 at 6:18 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/156#discussion_r39935568>:
>
> > @@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >        os << "input-dim=" << input_dim << " output-dim=" << output_dim;
> >        break;
> >      }
> > +    case 17: {
> > +      *component_type = "Convolutional1dComponent";
> > +      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4,
> > +	    patch_dim = 4 + Rand () % 5;
> > +
> > +      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
> > +      while ((patch_stride - patch_dim) % patch_step != 0)
>
> Done. One issue: I left the previous 2 commits (by david and vijay) in the
> interactive editor when using rebase, so currently there are 3 commits in
> this pull request. But actually I didn't modify anything in their commits.
> I am not sure if it is fine for you to merge only mine. If not, I can
> create a new pull request. Sorry about it.
> Yiming
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/156/files#r39935568>.
>

freewym replied:
I fixed that. Now there is only one commit of mine. 
By a test example, did you mean a NN with Convolutional1dComponent, rather than only testing a component? Could you please refer me to a similar example in nnet-test-utils.cc or somewhere else?
Thanks,
Yiming

danpovey replied:
See GenerateConfigSequence and the functions it calls.



On Sun, Sep 20, 2015 at 6:59 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/156#discussion_r39936061>:
>
> > @@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >        os << "input-dim=" << input_dim << " output-dim=" << output_dim;
> >        break;
> >      }
> > +    case 17: {
> > +      *component_type = "Convolutional1dComponent";
> > +      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4,
> > +	    patch_dim = 4 + Rand () % 5;
> > +
> > +      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
> > +      while ((patch_stride - patch_dim) % patch_step != 0)
>
> I fixed that. Now there is only one commit of mine.
> By a test example, did you mean a NN with Convolutional1dComponent, rather
> than only testing a component? Could you please refer me to a similar
> example in nnet-test-utils.cc or somewhere else?
> Thanks,
> Yiming
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/156/files#r39936061>.
>

danpovey replied:
You can add that to this pull request.

On Sun, Sep 20, 2015 at 7:58 PM, Daniel Povey <dpovey@gmail.com> wrote:

> See GenerateConfigSequence and the functions it calls.
>
>
>
> On Sun, Sep 20, 2015 at 6:59 PM, Yiming Wang <notifications@github.com>
> wrote:
>
>> In src/nnet3/nnet-test-utils.cc
>> <https://github.com/kaldi-asr/kaldi/pull/156#discussion_r39936061>:
>>
>> > @@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
>> >        os << "input-dim=" << input_dim << " output-dim=" << output_dim;
>> >        break;
>> >      }
>> > +    case 17: {
>> > +      *component_type = "Convolutional1dComponent";
>> > +      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4,
>> > +	    patch_dim = 4 + Rand () % 5;
>> > +
>> > +      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
>> > +      while ((patch_stride - patch_dim) % patch_step != 0)
>>
>> I fixed that. Now there is only one commit of mine.
>> By a test example, did you mean a NN with Convolutional1dComponent,
>> rather than only testing a component? Could you please refer me to a
>> similar example in nnet-test-utils.cc or somewhere else?
>> Thanks,
>> Yiming
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/156/files#r39936061>.
>>
>
>

danpovey commented:
Thanks!  Merging.
I believe in typical convolutional setups there is some kind pooling component (e.g. max-pooling) after each convolutional component.  Perhaps you could look into this, and port any max-pooling code to nnet3 (if necessary), and modify the test config accordingly?
